src/neptune_tensorboard/sync/sync_impl.py

```diff
@@ -0,0 +1,97 @@
+import hashlib
+import os
+import traceback
+
+import click
+import neptune
+from tensorboard.backend.event_processing.event_accumulator import EventAccumulator
+
+try:
+    import tbparse
+except ModuleNotFoundError:
+    raise ModuleNotFoundError("neptune-tensorboard: require `tbparse` for exporting logs (pip install tbparse)")
+
+
+class DataSync(object):
+    def __init__(self, project, api_token, path):
+        self._project = project
+        self._api_token = api_token
+        self._path = path
+
+    def run(self):
+        # Inspect if files correspond to EventFiles.
+        for root, _, run_files in os.walk(self._path):
+            for run_file in run_files:
+                try:
+                    path = os.path.join(root, run_file)
+                    # Skip if not a valid file
+                    if not self._is_valid_tf_event_file(path):
+                        continue
+                    self._export_to_neptune_run(path)
+                except Exception as e:
+                    click.echo("Cannot load run from file '{}'. ".format(run_file) + "Error: " + str(e))
+                    try:
+                        traceback.print_exc(e)
+                    except:  # noqa: E722
+                        pass
+
+    def _is_valid_tf_event_file(self, path):
+        accumulator = EventAccumulator(path)
+        accumulator.Reload()
+        try:
+            accumulator.FirstEventTimestamp()
+        except ValueError:
+            return False
+        return True
+
+    def _get_existing_neptune_run_ids(self):
+        with neptune.init_project(project=self._project, api_token=self._api_token) as project:
+            try:
+                existing_neptune_run_ids = {
+                    run_id for run_id in project.fetch_runs_table().to_pandas()["sys/custom_run_id"].to_list()
+                }
+            except KeyError:
+                # empty project
+                existing_neptune_run_ids = set()
+
+            return existing_neptune_run_ids
+
+    def _experiment_exists(self, hash_run_id, run_path):
+        existing_custom_ids = self._get_existing_neptune_run_ids()
```

Comment:
This is a "heavy" operation so consider doing it only once per project and as soon as possible (probably at the very beginning of `run` method)